### ISSUE TITLE
OOM issue when running ESLint command via Nx

### ISSUE THREAD

**Issue by akwodkiewicz** [nrwl/nx#27466]
### Current Behavior

ESLint is unable to finish linting the project due to OOM error

### Expected Behavior

ESLint finishes properly

### Steps to Reproduce

1. run `nx lint my-project` where `lint` is defined in `package.json` as `"run --top-level eslint src"`

I cannot share the contents of a private repository.

### Nx Report

```shell
Node   : 18.19.1
OS     : darwin-arm64
yarn   : 4.2.2

nx (global)  : 19.5.3
nx           : 19.5.3
@nrwl/tao    : 19.5.3
typescript   : 5.2.2
```

### Failure Logs

```shell
> nx run my-project:lint


<--- Last few GCs --->

[11044:0x128008000]   268782 ms: Scavenge 3758.3 (4123.5) -> 3747.1 (4124.8) MB, 6.7 / 0.0 ms  (average mu = 0.218, current mu = 0.248) allocation failure;
[11044:0x128008000]   268838 ms: Scavenge 3761.6 (4125.2) -> 3750.1 (4127.2) MB, 9.1 / 0.0 ms  (average mu = 0.218, current mu = 0.248) allocation failure;
[11044:0x128008000]   268909 ms: Scavenge 3763.8 (4127.2) -> 3750.8 (4127.9) MB, 15.2 / 0.0 ms  (average mu = 0.218, current mu = 0.248) allocation failure;


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x101069fe0 node::Abort() [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 2: 0x10106a1c4 node::ModifyCodeGenerationFromStrings(v8::Local<v8::Context>, v8::Local<v8::Value>, bool) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 3: 0x1011c073c v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 4: 0x10136b020 v8::internal::EmbedderStackStateScope::EmbedderStackStateScope(v8::internal::Heap*, v8::internal::EmbedderStackStateScope::Origin, cppgc::EmbedderStackState) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 5: 0x10136ec0c v8::internal::Heap::CollectSharedGarbage(v8::internal::GarbageCollectionReason) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 6: 0x10136bc20 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*, v8::GCCallbackFlags) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 7: 0x101368f20 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 8: 0x10135da5c v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
 9: 0x10135e2a0 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
10: 0x101344220 v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
11: 0x1016e4940 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
12: 0x101a3104c Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
13: 0x108c34d54
14: 0x108bce564
15: 0x108eca498
16: 0x107fa653c
17: 0x108eb3e74
18: 0x107a88ce8
19: 0x107a8ad0c
20: 0x108eb5b74
21: 0x101a467d8 Builtins_ArrayForEach [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
22: 0x1081af41c
23: 0x109209470
24: 0x108f2bbd8
25: 0x108daf704
26: 0x108b1dbe8
27: 0x1019bc198 Builtins_InterpreterEntryTrampoline [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
28: 0x1019bc198 Builtins_InterpreterEntryTrampoline [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
29: 0x1019edef4 Builtins_AsyncFunctionAwaitResolveClosure [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
30: 0x101a7c738 Builtins_PromiseFulfillReactionJob [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
31: 0x1019dfc4c Builtins_RunMicrotasks [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
32: 0x1019ba3a4 Builtins_JSRunMicrotasksEntry [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
33: 0x1012ece6c v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
34: 0x1012ed35c v8::internal::(anonymous namespace)::InvokeWithTryCatch(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
35: 0x1012ed538 v8::internal::Execution::TryRunMicrotasks(v8::internal::Isolate*, v8::internal::MicrotaskQueue*, v8::internal::MaybeHandle<v8::internal::Object>*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
36: 0x101313ca0 v8::internal::MicrotaskQueue::RunMicrotasks(v8::internal::Isolate*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
37: 0x10131443c v8::internal::MicrotaskQueue::PerformCheckpoint(v8::Isolate*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
38: 0x100facc4c node::InternalCallbackScope::Close() [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
39: 0x100facfd0 node::InternalMakeCallback(node::Environment*, v8::Local<v8::Object>, v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*, node::async_context) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
40: 0x100fc24a0 node::AsyncWrap::MakeCallback(v8::Local<v8::Function>, int, v8::Local<v8::Value>*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
41: 0x10106ee0c node::fs::FSReqCallback::Reject(v8::Local<v8::Value>) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
42: 0x10106f6a8 node::fs::FSReqAfterScope::Reject(uv_fs_s*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
43: 0x10106fac8 node::fs::AfterStat(uv_fs_s*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
44: 0x10106632c node::MakeLibuvRequestCallback<uv_fs_s, void (*)(uv_fs_s*)>::Wrapper(uv_fs_s*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
45: 0x101999174 uv__work_done [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
46: 0x10199c930 uv__async_io [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
47: 0x1019aeee4 uv__io_poll [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
48: 0x10199ce00 uv_run [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
49: 0x100fad6e0 node::SpinEventLoop(node::Environment*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
50: 0x1010a65dc node::NodeMainInstance::Run() [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
51: 0x1010364f0 node::LoadSnapshotDataAndRun(node::SnapshotData const**, node::InitializationResult const*) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
52: 0x1010367a8 node::Start(int, char**) [/opt/homebrew/Cellar/nvm/0.39.1/versions/node/v18.19.1/bin/node]
53: 0x18fd5e0e0 start [/usr/lib/dyld]
```

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

This is a medium-big project in TypeScript, utilizing `typescript-eslint` rules and referencing other internal packages.

When I run the command via Yarn, ESLint is able to finish its job in around 4 minutes.

```
hyperfine --runs 1 --ignore-failure 'yarn workspace my-project run lint'

Benchmark 1: yarn workspace my-project run lint
  Time (abs ≡):        258.713 s               [User: 442.243 s, System: 25.045 s]

  Warning: Ignoring non-zero exit code.
```

**Comment by jaysoo** [comment]
Can you share a repro on this?

**Comment by akwodkiewicz** [comment]
Unfortunately I cannot at the moment, because this is a private company monorepo, so for obvious reasons I cannot just make it public. And I still was not able to find the minimal set of things that trigger this issue.

The monorepo has dozens of projects, but the OOM issue hits only the biggest one (>10k files). The project references other packages, both using proper imports and using TS's path aliases. I have some refactors planned that will change the TypeScript setup a bit -- my gut feeling says that it might contribute to the issue.

Apologies for not being able to provide more info at this moment, but I decided to file a issue anyway, because this is bugging me for a long time and I figured out that maybe the maintainers have some quick ideas or maybe somebody else will be able to provide a repro under this ticket.

If I'm able to give more info, I'll definitely update you on this one.

**Comment by akwodkiewicz** [comment]
I kept the `parser: '@typescript-eslint/parser'` setting in the root `.eslintrc.yml` file (otherwise ESLint fails on parsing .ts files even for type-less rules) but removed all `@typescript-eslint/*` rules from the configuration file**s**.

It finished linting without OOM issues in 50s.

Before the changes ESLint would take around 4 minutes (if run without Nx, otherwise there was this OOM issue). So at least I know this _is_ somehow related to TypeScript and `@typescript-eslint` behaviour.

**Comment by akwodkiewicz** [comment]
Anyway, I don't think that Nx modifies ESLint's behaviour in any way, since it's run without any Nx plugins, so it feels like it should be only about resources management.

Maybe there's some other info I could provide, related to, for example, memory usage and we could operate on that?

**Comment by akwodkiewicz** [comment]
I was able to mitigate the issue by moving the `@typescript-eslint` rules from one file to another.

The affected project is configured by 3 `.eslintrc.yml` files. 2 files contain a `rules` key that includes various `@typescript-eslint` entries. I kept all of the files, but moved all `@typescript-eslint` rules from the 2nd config file to the root config file.

```
.eslintrc.yml   <-------------------------.
packages/              merged this with that
    .eslintrc.yml    <----------'  
    affected-project/
        .eslintrc.yml
```

I need to properly use `override` instead of simply moving the rules to `rules` (because now the rules from `packages` started affecting projects that are not located in `packages` (yes, there are projects like that). But at least I was able to quickly find a weird difference in tooling behaviour.
